Feature/314 enhance jenkins pipeline template to use updated package lifecycle (#326)
IBM/dbb: Templates/JenkinsPipeline/README.md

# Jenkins Pipeline template

> There are 2 Jenkins pipeline definition templates provided to set up a Jenkins Multibranch pipeline compatible with any Git provider. 

Reviewer: ```suggestion
Two Jenkins pipeline definition templates are provided to set up a Jenkins Multibranch pipeline compatible with any Git provider: 
```

> * a [Jenkinsfile](Jenkinsfile) definition to set up a Jenkins Multibranch pipeline using IBM UrbanCode Deoloy (renamed as IBM DevOps Deploy) as a deployment framework.

Reviewer: ```suggestion
* the [Jenkinsfile](Jenkinsfile) template is a pipeline implementation using IBM DevOps Deploy (formerly IBM UrbanCode Deploy) as a deployment framework.
```

> * a [JenkinsfileWithWaziDeploy](JenkinsfileWithWaziDeploy) definition to set up a Jenkins Multibranch pipeline using IBM Wazi Deploy as a deployment frameowork.

Reviewer: ```suggestion
* the [JenkinsfileWithWaziDeploy](JenkinsfileWithWaziDeploy) template uses IBM Wazi Deploy as the deployment framework.
```

## Overview and capabilities

> These two pipeline templates are implementing the [Git-based process and branching model for mainframe development](https://ibm.github.io/z-devops-acceleration-program/docs/branching/git-branching-model-for-mainframe-dev) with Jenkins as a the Pipeline Orchestrator.

Reviewer: ```suggestion
Both pipeline templates are implementing the [Git-based process and branching model for mainframe development](https://ibm.github.io/z-devops-acceleration-program/docs/branching/git-branching-model-for-mainframe-dev).
```

  * to publish deployment log files to the Jenkins build result,
  * to add links to the UCD Deployment request.
 * `Workspace Cleanup` stage to clean up the workspace.

 The [JenkinsfileWithWaziDeploy](JenkinsfileWithWaziDeploy) pipeline template implements the following stages:

* `Checkout` stage to clone the Git repository to a workspace directory on z/OS UNIX System Services using the integrated Git Plugin of Jenkins.
* `Pipeline Setup` to compute the settings for subsequent stages, that are displayed in the `Parameters and Values` stage.
* `Build` stage 
  * to invoke zAppBuild via the [dbbBuild.sh](../Common-Backend-Scripts/README.md#135---#dbbbuildsh-for-zappbuild-framework) Common Backend Script,
  * to publish log files to the Jenkins build result.
* `SonarQube Analysis` stage to request a SonarQube scan of the application repository.
* `Packaging` stage
  * to create a .TAR file package based on the [packageBuildOutputs.sh](https://github.com/IBM/dbb/blob/main/Templates/Common-Backend-Scripts/README.md#138---packageBuildOutputssh) script,

>   * to automatically compute the version of the next release for a release pipeline via [computeReleaseVersion.sh]((https://github.com/IBM/dbb/blob/main/Templates/Common-Backend-Scripts/README.md#137---computeReleaseVersionsh), 

Reviewer: Doesn't this have to be lsited before `packageBuildOutputs`?

  * to automatically compute the artifact repository URL to store the .TAR file package,
  * to upload the .TAR file package to the artifact repository on the computed URL.
* `Deploy Integration` stage to deploy to the development / integration test environment that includes:

>   * to run the Wazi Deploy and generate deployment plan [generate command](../Common-Backend-Scripts/README.md#140---wazideploy-generatesh),

Reviewer: ```suggestion
  * to run the [wazideploy-generate.sh script](../Common-Backend-Scripts/README.md#140---wazideploy-generatesh) to download the .TAR file package from the artifact repository URL and the generate deployment plan
```

>   * to download the .TAR file package from the artifact repository URL that is uploaded during the packaging stage,

Reviewer: ```suggestion
```

  * to deploy the .TAR file package with the Wazi Deploy [deploy command](../Common-Backend-Scripts/README.md#141---wazideploy-deploysh) (Python-based).

>   * to use Wazi Deploy evidence command [deploy command](../Common-Backend-Scripts/README.md#142---wazideploy-generatesh) to index deploy invormation for Deployment report from the Wazi Deploy Evidence YAML file.

Reviewer: ```suggestion
  * to execute the [Wazi Deploy evidence script](../Common-Backend-Scripts/README.md#142---wazideploy-generatesh) to generate a deployment report that is attached to the pipeline run and to push evidences into a shared directory structure for further analysis of deployment results.
```
